plugin/keyring_vault/vault_parser.cc

```diff
@@ -0,0 +1,178 @@
+#include "vault_parser.h"
+#include "vault_key.h"
+#include <vector>
+#include <sstream>
+#include <algorithm>
+#include "vault_base64.h"
+
+namespace keyring
+{
+
+bool Vault_parser::retrieve_tag_value(const Secure_string &payload, const Secure_string &tag, const char opening_bracket,
+                                      const char closing_bracket, Secure_string *value)
+{
+  value->clear();
+  size_t opening_bracket_pos, closing_bracket_pos, tag_pos = payload.find(tag);
+  if (tag_pos == Secure_string::npos)
+    return false;
+
+  if ((opening_bracket_pos = (payload.find(opening_bracket, tag_pos))) == Secure_string::npos ||
+      (closing_bracket_pos = (payload.find(closing_bracket, opening_bracket_pos))) == Secure_string::npos)
+  {
+    std::ostringstream err_ss("Could not parse tag ");
+    err_ss << tag << " from Vault's response."; 
+    logger->log(MY_ERROR_LEVEL, err_ss.str().c_str());
+    return true;
+  }
+
+ *value = payload.substr(opening_bracket_pos, closing_bracket_pos - opening_bracket_pos +1);
+  value->erase(std::remove(value->begin(), value->end(), '\n'), value->end());
+  return false;
+}
+
+bool Vault_parser::retrieve_list(const Secure_string &payload, const Secure_string &list_name, Secure_string *list)
+{
+  return retrieve_tag_value(payload, list_name, '[', ']', list);
+}
+
+bool Vault_parser::retrieve_map(const Secure_string &payload, const Secure_string &map_name, Secure_string *map)
+{
+  return retrieve_tag_value(payload, map_name, '{', '}', map);
+}
+
+bool Vault_parser::retrieve_tokens_from_list(const Secure_string &list, std::vector<Secure_string> *tokens)
+{
+  std::size_t token_start = 0, token_end = 0;
+  while ((token_start = list.find('\"', token_end))
+           != Secure_string::npos &&
+         token_start < list.size())
+  {
+    if ((token_end = list.find('\"', token_start+1)) == Secure_string::npos)
+    {
+      tokens->clear();
+      return true;
+    }
+    tokens->push_back(list.substr(token_start+1,
+                                   token_end-token_start-1));
+    ++token_end;
+  }
+  return false;
+}
+
+bool Vault_parser::retrieve_value_from_map(const Secure_string &map, const Secure_string &key,
+                                           Secure_string *value)
+{
+  size_t key_tag_pos = Secure_string::npos, value_start_pos = Secure_string::npos, 
+         value_end_pos = Secure_string::npos;
+  bool was_error = false;
+
+  if ((key_tag_pos = map.find(key)) != Secure_string::npos &&
+      (value_start_pos = map.find(":\"", key_tag_pos)) != Secure_string::npos &&
```

Comment:
I find it quite unlikely as spaces can be valid values for values stored in JSON. If they would allow random spaces between tags, how would they tell if those are not spaces that belongs to the value.